### PR TITLE
`ipfs get` Command

### PR DESCRIPTION
**Pull request by mappum**
(This PR should be merged after #613)

This adds the `ipfs get` command, which downloads unixfs files to disk. It differs from cat in that it can download directories, and it outputs to a file path instead of Stdout.

It supports some options:
`--output` or `-o`: the path where the output files should be saved
`--archive` or `-a`: output a TAR archive instead of individual files
`--compress` or `-c`: GZIP the output (useful when downloading from a remote daemon, or use with `-a` if you want a .tar.gz.
`--compression-level` or `-l`: The level of compression (1-9)

### CODE REVIEW

**Comment by whyrusleeping** [comment]
I wish that it were easy to hide vendored code from PRs... my chromebook hates loading and rendering it all

**Comment by jbenet** [comment]
pro tip: use the compare view: 535d52a...32f9226  (the rest is other PRs)

File: core/commands/get.go [hunk 1]
```diff
@@ -0,0 +1,426 @@
...
+	},
+
+	Arguments: []cmds.Argument{
+		cmds.StringArg("ipfs-path", true, false, "The path to the IPFS object(s) to be outputted").EnableStdin(),
+	},
+	Options: []cmds.Option{
+		cmds.StringOption("output", "o", "The path where output should be stored"),
+		cmds.BoolOption("archive", "a", "Output a TAR archive"),
+		cmds.BoolOption("compress", "C", "Compress the output with GZIP compression"),
+		cmds.IntOption("compression-level", "l", "The level of compression (an int between 1 and 9)"),
```
Comment (jbenet) [inline core/commands/get.go]: maybe just: `"The level of compression (1-9)"`

File: core/commands/get.go [hunk 2]
```diff
@@ -0,0 +1,426 @@
...
+				compressionLevel = gzip.DefaultCompression
+			} else {
+				compressionLevel = gzip.NoCompression
+			}
+		} else {
+			if compressionLevel < 1 || compressionLevel > 9 {
+				res.SetError(ErrInvalidCompressionLevel, cmds.ErrClient)
+				return
+			}
+		}
```
Comment (jbenet) [inline core/commands/get.go]: switches make these flows a bit easier to parse:

```go
cmprs, _, _ := req.Option("compress").Bool()
cmplvl, cmplvlFound, _ := req.Option("compression-level").Int()
switch {
case !cmprs:
	cmplvl = gzip.NoCompression
case cmprs && !cmplvlFound:
	cmplvl = gzip.DefaultCompression
case cmprs && cmplvlFound && (cmplvl < 1 || cmplvl > 9):
	res.SetError(ErrInvalidCompressionLevel, cmds.ErrClient)
		return
}
```

File: core/commands/get.go [hunk 3]
```diff
@@ -0,0 +1,426 @@
...
+		res.SetOutput(nil)
+
+		outPath, _, _ := req.Option("output").String()
+		if len(outPath) == 0 {
+			outPath = req.Arguments()[0]
+		}
+
+		compress, _, _ := req.Option("compress").Bool()
+		compressionLevel, found, _ := req.Option("compression-level").Int()
+		compress = (compress && (compressionLevel > 0 || !found)) || compressionLevel > 0
```
Comment (jbenet) [inline core/commands/get.go]: a bit hard to follow.
```
cmprs, _, _ := req.Option("compress").Bool()
cmplvl, cmplvlFound, _ := req.Option("compression-level").Int()
if !cmprs {
    cmprs = cmplvl > 0
}
```

File: core/commands/get.go [hunk 4]
```diff
@@ -0,0 +1,426 @@
...
+				res.SetError(err, cmds.ErrNormal)
+				return
+			}
+
+			err = file.Close()
+			if err != nil {
+				res.SetError(err, cmds.ErrNormal)
+				return
+			}
+		}
```
Comment (jbenet) [inline core/commands/get.go]: a ton of this can be broken out into util funcs in small pkgs inside `thirdparty/`. For example:

```go

type TarExpander struct {
  Path string
}

func (te *TarExpander) Expand(r tar.Reader) error {}
func (te *TarExpander) ExpandDir(h tar.Header) error {}
func (te *TarExpander) ExpandFile(h tar.Header, r tar.Reader) error {}
```

**Comment by jbenet** [comment]
Hard to follow the rest-- let's get #613 merged in

File: test/sharness/t0090-get.sh [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+  echo "Saving archive to "$HASH2".tar.gz
+" > expected &&
+  test_cmp expected actual
+'
+
+test_expect_success "gzipped tar archive output is valid (directory)" '
+  tar -xf "$HASH2".tar.gz &&
+  test_cmp dir/a "$HASH2"/a &&
+  test_cmp dir/b/c "$HASH2"/b/c &&
+  rm -r $HASH2
```
Comment (jbenet) [inline test/sharness/t0090-get.sh]: Awesome tests!!

**Comment by jbenet** [comment]
@mappum LGTM
